File: innerlibs1/FontAwesome.vb
Imports System.IO
Imports System.Runtime.CompilerServices

Public Module FontAwesome


    ''' <summary>
    ''' Retorna a classe do icone do FontAwesome que representa melhor o arquivo ou diretório
    ''' </summary>
    ''' <param name="File">Arquivo</param>
    ''' <returns></returns>
    <Extension>
    Public Function GetIconByFileType(File As FileSystemInfo, Optional DirectoryOpen As Boolean = False, Optional InvertIcon As Boolean = False) As String
        If File.Attributes = FileAttributes.Device Then
            Return "fa-plug"
        End If
        If File.Attributes = FileAttributes.Directory Then
            Select Case True
                Case DirectoryOpen And InvertIcon
                    Return "fa-folder-open-o"
                Case DirectoryOpen And Not InvertIcon
                    Return "fa-folder-open"
                Case Not DirectoryOpen And InvertIcon
                    Return "fa-folder-o"
                Case Else
                    Return "fa-folder"
            End Select
        Else
            Return File.Extension.GetIconByFileExtension()
        End If
    End Function

    ''' <summary>
    ''' Retorna a classe do icone do FontAwesome que representa melhor o arquivo
    ''' </summary>
    ''' <param name="MIME">MIME Type do Arquivo</param>
    ''' <returns></returns>
    <Extension()>
    Public Function GetIconByFileType(MIME As FileType) As String
        Return MIME.Extensions.FirstOr("").GetIconByFileExtension()
    End Function

    ''' <summary>
    ''' Retorna a classe do icone do FontAwesome que representa melhor o arquivo
    ''' </summary>
    ''' <param name="Extension">Arquivo</param>
    ''' <returns></returns>
    <Extension()>
    Public Function GetIconByFileExtension(Extension As String) As String
        Select Case Extension.RemoveAny(".").ToLower
            Case "png", "jpg", "gif", "jpeg", "psd", "ai", "drw", "ttf", "svg", "eps", "tiff", "cdr"
                Return "fa-file-picture-o"
            Case "doc", "docx"
                Return "fa-file-word-o"
            Case "pdf"
                Return "fa-file-pdf-o"
            Case "ppt", "pptx"
                Return "fa-file-powerpoint-o"
            Case "xls", "xlsx"
                Return "fa-file-excel-o"
            Case "html", "htm", "php", "cpp", "vb", "cs", "jsp", "xml", "js", "css", "aspx", "ascx", "ashx", "config", "ps1"
                Return "fa-file-code-o"
            Case "apk"
                Return "fa-android"
            Case "ios", "ipa"
                Return "fa-apple"
            Case "xap", "appx"
                Return "fa-windows"
            Case "zip", "rar", "tar", "jar", "gz", "iso", "7zip", "b1", "bar", "rar5", "pk3", "pkg"
                Return "fa-file-archive-o"
            Case "avi", "mpeg", "mp4", "3gp", "mkv", "wmv", "rmvb", "mov", "webm", "ogv"
                Return "fa-file-video-o"
            Case "txt", "otf", "otd", "ttf", "rtf", "csv", "xps", "cfg"
                Return "fa-file-text-o"
            Case "mp3", "mp2", "wma", "wav", "ogg", "flac", "aac"
                Return "fa-file-audio-o"
            Case "gb", "gba", "n64", "rom", "z64", "gbc", "wad"
                Return "fa-gamepad"
            Case "bin", "dll"
                Return "fa-cog"
            Case "exe", "bat", "msi"
                Return "fa-window-maximize"
            Case "sql"
                Return "fa-database"
            Case Else
        End Select
        Return "fa-file-o"
    End Function

    Public Const CDNFontAwesomeCSS = "https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"

End Module

File: InnerLibs/FontAwesome.cs
using InnerLibs.LINQ;
using System.IO;

namespace InnerLibs.FontAwesome
{
    public static class FontAwesome
    {
        /// <summary>
        /// Retorna a classe do icone do FontAwesome que representa melhor o arquivo
        /// </summary>
        /// <param name="Extension">Arquivo</param>
        /// <returns></returns>
        public static string GetIconByFileExtension(this string Extension)
        {
            switch (Extension.RemoveAny(".").ToLower() ?? "")
            {
                case "vcf":
                case "vcard":
                    {
                        return "fa-address-card";
                    }
                case "ics":
                case "ical":
                case "ifb":
                case "icalendar":
                    {
                        return "fa-calendar";
                    }
                case "png":
                case "jpg":
                case "gif":
                case "jpeg":
                case "psd":
                case "ai":
                case "drw":
                case "svg":
                case "eps":
                case "tiff":
                case "webp":
                case "cdr":
                    {
                        return "fa-file-image";
                    }

                case "doc":
                case "docx":
                    {
                        return "fa-file-word";
                    }

                case "pdf":
                    {
                        return "fa-file-pdf";
                    }

                case "ppt":
                case "pptx":
                    {
                        return "fa-file-powerpoint";
                    }

                case "xls":
                case "xlsx":
                    {
                        return "fa-file-excel";
                    }

                case "html":
                case "htm":
                case "php":
                case "cpp":
                case "vb":
                case "cs":
                case "jsp":
                case "xml":
                case "css":
                case "aspx":
                case "ascx":
                case "ashx":
                case "config":
                case "json":
                case "jsx":
                case "js":
                case "ts":
                case "vbs":
                case "ps1":
                    {
                        return "fa-file-code";
                    }
                case "apk":
                case "appbundle":
                    {
                        return "fa-android";
                    }

                case "ios":
                case "ipa":
                    {
                        return "fa-apple";
                    }

                case "xap":
                case "appx":
                    {
                        return "fa-windows";
                    }

                case "zip":
                case "rar":
                case "tar":
                case "gz":
                case "7zip":
                case "7z":
                case "b1":
                case "bar":
                case "rar5":
                case "pk3":
                case "pkg":
                    {
                        return "fa-file-archive";
                    }

                case "avi":
                case "mpeg":
                case "mp4":
                case "3gp":
                case "mkv":
                case "wmv":
                case "rmvb":
                case "mov":
                case "webm":
                case "ogv":
                    {
                        return "fa-file-video";
                    }

                case "txt":
                case "otf":
                case "otd":
                case "ttf":
                case "rtf":
                case "xps":
                case "cfg":
                    {
                        return "fa-file-text";
                    }

                case "csv":
                    {
                        return "fa-file-csv";
                    }

                case "mp3":
                case "mp2":
                case "wma":
                case "wav":
                case "ogg":
                case "flac":
                case "aac":
                    {
                        return "fa-file-audio";
                    }

                case "gb":
                case "gba":
                case "n64":
                case "rom":
                case "z64":
                case "gbc":
                case "smc":
                case "sfc":
                case "wad":
                case "ndc":
                case "gci":
                case "3ds":
                case "nes":
                case "snes":
                case "cia":
                case "gcz":
                    {
                        return "fa-gamepad";
                    }
                case "iso":
                case "ape":
                case "bwt":
                case "ccd":
                case "cdi":
                case "cue":
                case "b5t":
                case "b6t":
                    {
                        return "fa-compact-disc";
                    }

                case "dll":
                    {
                        return "fa-cog";
                    }

                case "exe":
                case "bat":
                case "msi":
                    {
                        return "fa-window-maximize";
                    }

                case "sql":
                case "db":
                case "sqlite":
                case "litedb":
                case "mdb":
                case "mdf":
                    {
                        return "fa-database";
                    }
                case "bak":
                    {
                        return "fa-copy";
                    }
                case "jar":
                    {
                        return "fa-java";
                    }

                default:
                    {
                        break;
                    }
            }

            return "fa-file";
        }

        /// <summary>
        /// Retorna a classe do icone do FontAwesome que representa melhor o arquivo ou diretório
        /// </summary>
        /// <param name="File">Arquivo</param>
        /// <returns></returns>
        public static string GetIconByFileType(this FileSystemInfo File, bool DirectoryOpen = false)
        {
            if (File.Attributes == FileAttributes.Device)
            {
                return "fa-plug";
            }
            else if (File.Attributes == FileAttributes.Directory)
            {
                return DirectoryOpen ? "fa-folder-open" : "fa-folder";
            }
            else
            {
                return File.Extension.GetIconByFileExtension();
            }
        }

        /// <summary>
        /// Retorna a classe do icone do FontAwesome que representa melhor o arquivo
        /// </summary>
        /// <param name="MIME">MIME Type do Arquivo</param>
        /// <returns></returns>
        public static string GetIconByFileType(this FileType MIME) => MIME.Extensions.FirstOr("").GetIconByFileExtension();
    }
}